### PR TITLE
Shooping cart implemented:

### PR DESCRIPTION
**Pull request by BohdanS-UA**
Liquibase: shopping-cart-table, cart-item-table schemas created

ShoppingCartController, ShoppingCartService, ShoppingCartRepository, ShoppingCartMapper added

ShoppingCart, ShoppingCartDto,added

CartItem, CartItemAddBookResponseDto, CartItemDto, CartItemUpdateDto added

CartItemRepository added

### CODE REVIEW

File: src/main/java/mate/academy/bookstorev2/controller/ShoppingCartController.java [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Shopping Cart management", description = "Endpoints for managing shopping carts")
+@RestController
+@RequiredArgsConstructor
+@RequestMapping("/api/cart")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/controller/ShoppingCartController.java]:
```suggestion
@RequestMapping("/cart")
```

add `server.servlet.context-path=/api` to application.properties

File: src/main/java/mate/academy/bookstorev2/controller/ShoppingCartController.java [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+            description = "Get information about shopping cart")
+    public ShoppingCartDto getCart(@AuthenticationPrincipal User user,
+                                   @ParameterObject @PageableDefault Pageable pageable) {
+        return shoppingCartService.getShoppingCart(user, pageable);
+    }
+
+    @PutMapping("/items/{cartItemId}")
+    @Operation(summary = "Update cart item",
+            description = "Update quantity in cart item by cart item's id")
+    public void updateCartItem(@PathVariable @Positive Long cartItemId,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto updateCartItem(@PathVariable @Positive Long cartItemId,
```

File: src/main/java/mate/academy/bookstorev2/controller/ShoppingCartController.java [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+import org.springframework.web.bind.annotation.GetMapping;
+import org.springframework.web.bind.annotation.PathVariable;
+import org.springframework.web.bind.annotation.PostMapping;
+import org.springframework.web.bind.annotation.PutMapping;
+import org.springframework.web.bind.annotation.RequestBody;
+import org.springframework.web.bind.annotation.RequestMapping;
+import org.springframework.web.bind.annotation.RestController;
+
+@Tag(name = "Shopping Cart management", description = "Endpoints for managing shopping carts")
+@RestController
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/controller/ShoppingCartController.java:28]:
```suggestion
@RestController
@Validated
```

File: src/main/java/mate/academy/bookstorev2/controller/ShoppingCartController.java [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+@RequiredArgsConstructor
+@RequestMapping("/api/cart")
+public class ShoppingCartController {
+    private final ShoppingCartService shoppingCartService;
+
+    @PostMapping
+    @Operation(summary = "Add a book to the shopping cart",
+            description = "Add a book to the shopping cart")
+    public CartItemAddBookResponseDto addBookToShoppingCart(
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/controller/ShoppingCartController.java]:
```suggestion
    public ShoppingCartDto addBookToShoppingCart(
```

File: src/main/java/mate/academy/bookstorev2/controller/ShoppingCartController.java [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+            @RequestBody @Valid AddToShoppingCartDto dto,
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/controller/ShoppingCartController.java]:
```suggestion
            @RequestBody @Valid CartItemRequestDto cartItemDto,
```

File: src/main/java/mate/academy/bookstorev2/dto/cart/item/CartItemUpdateDto.java [hunk 6]
```diff
@@ -0,0 +1,11 @@
+package mate.academy.bookstorev2.dto.cart.item;
+
+import jakarta.validation.constraints.Positive;
+import org.springframework.lang.NonNull;
+
+public record CartItemUpdateDto(
+        @Positive
+        @NonNull
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/dto/cart/item/CartItemUpdateDto.java]:
```suggestion
```

int can't be null

File: src/main/java/mate/academy/bookstorev2/model/ShoppingCart.java [hunk 7]
```diff
@@ -0,0 +1,37 @@
...
+import lombok.Setter;
+import org.hibernate.annotations.SQLDelete;
+import org.hibernate.annotations.Where;
+
+@Entity
+@Getter
+@Setter
+@SQLDelete(sql = "UPDATE shopping_cart SET is_deleted=true WHERE id=?")
+@Where(clause = "is_deleted=false")
+@Table(name = "shopping_cart")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/model/ShoppingCart.java]:
```suggestion
@Table(name = "shopping_carts")
```

File: src/main/java/mate/academy/bookstorev2/model/CartItem.java [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+import jakarta.persistence.JoinColumn;
+import jakarta.persistence.ManyToOne;
+import jakarta.persistence.Table;
+import lombok.Getter;
+import lombok.Setter;
+
+@Entity
+@Getter
+@Setter
+@Table(name = "cart_item")
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/model/CartItem.java]:
```suggestion
@Table(name = "cart_items")
```

File: src/main/java/mate/academy/bookstorev2/model/User.java [hunk 9]
```diff
@@ -41,6 +43,8 @@ public class User implements UserDetails {
             inverseJoinColumns = @JoinColumn(name = "role_id")
     )
     private Set<Role> roles = new HashSet<>();
+    @OneToOne(mappedBy = "user", cascade = CascadeType.ALL, orphanRemoval = true)
+    private ShoppingCart shoppingCart;
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/model/User.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstorev2/service/ShoppingCartServiceImpl.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    private final CartItemsMapper cartItemsMapper;
+
+    @Override
+    public ShoppingCart createShoppingCartForUser(User user) {
+        ShoppingCart cart = new ShoppingCart();
+        cart.setUser(user);
+        return shoppingCartRepository.save(cart);
+    }
+
+    @Override
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/service/ShoppingCartServiceImpl.java:35]: Check where is Transactional needed

File: src/main/java/mate/academy/bookstorev2/service/ShoppingCartServiceImpl.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    public void updateCartItem(Long cartItemId, CartItemUpdateDto cartItemUpdateDto) {
+        CartItem cartItem = cartItemRepository.findById(cartItemId)
+                .orElseThrow(() -> new EntityNotFoundException("Cart item not found"));
+        cartItem.setQuantity(cartItemUpdateDto.quantity());
+        cartItemRepository.save(cartItem);
+    }
+
+    @Override
+    @Transactional
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/service/ShoppingCartServiceImpl.java]:
```suggestion
```

File: src/main/java/mate/academy/bookstorev2/service/ShoppingCartServiceImpl.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+    public ShoppingCart createShoppingCartForUser(User user) {
+        ShoppingCart cart = new ShoppingCart();
+        cart.setUser(user);
+        return shoppingCartRepository.save(cart);
+    }
+
+    @Override
+    public CartItemAddBookResponseDto addToCart(User user, AddToShoppingCartDto dto) {
+        ShoppingCart cart = shoppingCartRepository.findByUserId(user.getId())
+                .orElseGet(() -> createShoppingCartForUser(user));
```
Comment (Elena-Bruyako) [inline src/main/java/mate/academy/bookstorev2/service/ShoppingCartServiceImpl.java]: you create shop cart during the registration, if you can't find shop cart for user - throw exception

**Review by Elena-Bruyako — CHANGES_REQUESTED** [review]
see comments

**Review by sarakhmen — APPROVED** [review]

**Pull request merged**